Feat: Add a multi-select on relation filters

AyushAgrawal-A2 opened the pull request:
Closes #1719

- Create multi-select filters with operand select for the user, people & company entity.



https://github.com/twentyhq/twenty/assets/54364088/57f34179-3027-4b69-882b-6aaecf6ac800

AyushAgrawal-A2 commented on front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:
```diff
@@ -0,0 +1,128 @@
+    );
+
+  const [selectedEntityIds, setSelectedEntityIds] = useState<
+    Record<string, { name: string; isChecked: boolean }>
```
done

lucasbordeau commented:
I'm not able to see the changes you made, could you merge main and push ?

charlesBochet commented:
@Bonapara does design looks good to you? :)

AyushAgrawal-A2 commented:
> I'm not able to see the changes you made, could you merge main and push ?

Found a bug while screen recording. Just fixed it.

lucasbordeau commented on front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:
```diff
@@ -0,0 +1,113 @@
+  const removeFilter = useRemoveFilter();
+
+  const handleChange = (enitityId: string, newCheckedValue: boolean) => {
+    setSelectedEntities((prev) =>
```
We avoid generic names even in lambdas, better to have `previousSelectedEntities`

lucasbordeau commented on front/src/modules/ui/view-bar/components/FilterDropdownMultipleEntitySearchSelect.tsx:
```diff
@@ -0,0 +1,113 @@
+  const handleChange = (enitityId: string, newCheckedValue: boolean) => {
+    setSelectedEntities((prev) =>
+      prev.map((entity) => {
+        if (entity.id === enitityId) {
```
`enitity`

lucasbordeau commented:
@AyushAgrawal-A2 

- Can you see if `OverflowingTextWithTooltip` can fit your needs ? Using this component + a max-width on the chip ?
- The API is way cleaner, nice !
- Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
- We should erase the selected values when closing the dropdown. => Meaning we should reset the state of the multiple selection so we don't have the previous selection shown in the dropdown, but it seems to be fixed with your new changes !

AyushAgrawal-A2 commented:
> * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.

I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it.
Can revert back if that is the desired behavior.

Bonapara commented:
> @Bonapara
> 
> 1. I have set the display name for multi-select filter as a comma-separated list. This will get too long.
>    ![image](https://user-images.githubusercontent.com/54364088/272644376-661591b0-3a65-4b43-8248-504220cc0673.png)
>    Options:
>    
>    * Trim it after a max length "Facebook, Google, Air..."
>    * Show full name if only one is selected and number if many e.g. "3 Selected"
>    * ??
> 2. Operand name in Operand Select menu
>    ![image](https://user-images.githubusercontent.com/54364088/272645467-d659ef2a-0058-4df0-b829-455d057edb1b.png)
> 3. Operand in Filter chip. "Company: NotIn ..." i.e. `: In` `: NotIn` part. I feel `: NotIn` is not readable.

Hi @AyushAgrawal-A2 

- For your first question, let's do you second suggestion → Show full name if only one is selected and number if many e.g. "3 Selected"
- For your second question I think we can keep the behavior of people operand name : "is" and "not" (without displaying "is" in the chip)

lucasbordeau commented:
> > * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
> 
> I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it. Can revert back if that is the desired behavior.

Good point, what are your thoughts on that @Bonapara, I feel like it's better to show changes on each click.

AyushAgrawal-A2 commented:
@lucasbordeau @Bonapara 
- Made all changes in "as per feedback" commit [fcb51a6](https://github.com/twentyhq/twenty/pull/1822/commits/fcb51a6e122cf57ba0473ce16b89ab96dde95fc3)
- Switched to per click update in last commit [364809e](https://github.com/twentyhq/twenty/pull/1822/commits/364809e114fa15e2629294e5be2daa522acb28cc)

Please try it and decide.

P.S.: storybook fix is pending, and can be fixed only after behavior is finalized

Bonapara commented:
> > > * Is it difficult for you to change the filter at every select/unselect, not after we exit the dropdown ? We can always do it in another ticket.
> > 
> > 
> > I actually first implemented per-click update. On every selection, the table was getting refreshed and the menu item order was also getting changed as the selected item would move to the top. I didn't like the user experience while testing it. Can revert back if that is the desired behavior.
> 
> Good point, what are your thoughts on that @Bonapara, I feel like it's better to show changes on each click.

I agree with your understanding. Implementing changes on each click could enhance the experience, but we should prevent the menu from reordering after each selection. The reordering should only occur when the dropdown is closed and then reopened. We should open a new issue for that.

lucasbordeau commented:
I'll let @charlesBochet review because I'm quite busy but seems ok for me !

AyushAgrawal-A2 commented:
@Bonapara 

> but we should prevent the menu from reordering after each selection. The reordering should only occur when the dropdown is closed and then reopened. We should open a new issue for that.

This will require refactoring and reviewing the same code again, Please let me fix this in this PR itself.

AyushAgrawal-A2 commented:
@lucasbordeau @charlesBochet @Bonapara 
Changes:-
- Fixed menu item reorder on item select/deselect.
- Fixed correct operand select on filter menu open.


https://github.com/twentyhq/twenty/assets/54364088/e53310ef-5240-4985-aa64-5d723b8cc483

Submitting again for review.

lucasbordeau commented:
Waiting for relation field type refactor with new flexible backend.

lucasbordeau commented:
Hi @AyushAgrawal-A2, thank you very much for your efforts. 

After waiting for a long time, we have decided to implement this feature with the core team because it touches a lot of moving parts of the new architecture, in https://github.com/twentyhq/twenty/pull/2858.

Please feel free to tell us anything we might have missed in our implementation.
